projects/driver_controls/src/throttle.c

```diff
@@ -0,0 +1,133 @@
+#include "throttle.h"
+#include <string.h>
+#include "event_queue.h"
+#include "input_event.h"
+
+// Scales a reading to a measure out of 12 bits based on the given range (min - max).
+static uint16_t prv_scale_reading(int16_t reading, int16_t max, int16_t min) {
+  return (1 << 12) * (reading - min) / (max - min);
+}
+
+static uint16_t prv_get_numerator(int16_t reading, ThrottleZone zone, ThrottleChannel channel,
+                                  ThrottleStorage *storage) {
+  return prv_scale_reading(
+      reading, storage->calibration_data->zone_thresholds[channel][zone][THROTTLE_THRESH_MAX],
+      storage->calibration_data->zone_thresholds[channel][zone][THROTTLE_THRESH_MIN]);
+}
+
+static bool prv_reading_within_zone(int16_t reading, ThrottleZone zone, ThrottleChannel channel,
+                                    ThrottleStorage *storage) {
+  return (
+      (reading < storage->calibration_data->zone_thresholds[channel][zone][THROTTLE_THRESH_MAX]) &&
+      (reading >= storage->calibration_data->zone_thresholds[channel][zone][THROTTLE_THRESH_MIN]));
+}
+
+// Returns true if the channels match their supposed relationship.
+static bool prv_channels_synced(int16_t reading_main, int16_t reading_secondary, ThrottleZone zone,
+                                ThrottleStorage *storage) {
+  uint16_t numerator_main = prv_get_numerator(reading_main, zone, THROTTLE_CHANNEL_MAIN, storage);
+
+  uint16_t numerator_secondary =
+      prv_get_numerator(reading_secondary, zone, THROTTLE_CHANNEL_SECONDARY, storage);
+
+  return (abs(numerator_main - numerator_secondary)) <=
+         storage->calibration_data->channel_readings_tolerance;
+}
+
+// This callback is called whenever a conversion is done. It sets the flags
+// so that we know conversions are happening.
+static void prv_flag_update_callback(Ads1015Channel channel, void *context) {
+  ThrottleStorage *storage = context;
+  storage->reading_updated_flag = true;
+}
+
+// The periodic callback which checks if readings are up to date and channels are in sync.
+// If they are, the event corresponding to the pedals position is raised.
+// If not, a pedal timout event is raised.
+static void prv_raise_event_timer_callback(SoftTimerID timer_id, void *context) {
+  ThrottleStorage *storage = context;
+  int16_t reading_main = INT16_MIN;
+  int16_t reading_secondary = INT16_MIN;
+
+  ads1015_read_raw(storage->pedal_ads1015_storage, storage->channel_main, &reading_main);
+  ads1015_read_raw(storage->pedal_ads1015_storage, storage->channel_secondary, &reading_secondary);
+
+  InputEvent pedal_events[NUM_THROTTLE_ZONES] = { INPUT_EVENT_PEDAL_BRAKE, INPUT_EVENT_PEDAL_COAST,
+                                                  INPUT_EVENT_PEDAL_PRESSED };
+  bool fault = true;
+  if (storage->reading_updated_flag) {
+    for (ThrottleZone zone = THROTTLE_ZONE_BRAKE; zone < NUM_THROTTLE_ZONES; zone++) {
+      if (prv_reading_within_zone(reading_main, zone, THROTTLE_CHANNEL_MAIN, storage) &&
```

Comment:
What happens if the channels are reading different zones?

Reply:
I added a check for both readings being in the same zone inside the channel_synch function.

Reply:
Hmm, wouldn't that result in a failure if we're on the edge between two zones?
i.e. if the threshold for coast -> brake is `x`, it'd be possible for one channel to at be `x` while the second channel is `x` + tolerance. That would put them in different zones, but it'd still be valid.

Reply:
So for the function reading_within_zone maybe I should account for the tolerance

Reply:
Ah thats kinda tough since tolerance is not the same unit as readings or thresholds.
I have do it some other way

Reply:
I think it'd help if you fleshed out exactly how tolerance is calculated and where it's applied.

Reply:
My old approach with ZONE_ALL didnt have this problem because it wasnt checking the relationship between two channels based on a single zone.

Reply:
Sure. In that case, you don't even need zone information for the second channel.
I'd still like to see exactly how tolerance is calculated. I don't think your current approach really works.